### DISCUSSION TITLE
Issue with getting clean audio

### DISCUSSION THREAD

**Discussion by tnk-rin** [pschatzmann/ESP32-A2DP#593] [Q&A] [answered]
Hello!
Not too sure if this or the Issues is the best place to ask about this but I figure its best to ask since I've been wracking my brain on this for too long haha!
I've recently purchased an ESP32 and some PCM5102 DACs to build a BT receiver for my car, and wiring it up as per the examples gives 2 different results depending on which example is used. If I use the `bt_music_receiver_simple.ino` example, it will play audio for about half a second then freeze on the last several frames. If I use the ESP32 I2S API example in the Readme, it plays audio without freezing but instead lags every quarter or half second.
I then went to try the Espressif example, and got the same result with lag every half second or so. Using the serial monitor though yielded this:
\
![image](https://github.com/user-attachments/assets/657422bb-a49c-415d-9379-538a589b9398)
\
I'm not well versed in ring buffers so I'm not sure how this all works, but it seems like its having issues maintaining a buffer for the DAC. I'm using a Samsung Galaxy Note 9, if that's relevant, and the issue springs in both streamed and locally played music.

**Comment by tnk-rin** [comment]
Adding onto this, here's a screenshot of the waveform to give a better idea of what its doing.
![image](https://github.com/user-attachments/assets/116bb114-9123-40f3-90c6-3b19c2ec7a20)

**Comment by pschatzmann** [comment, marked as answer]
This does not make any sense to me: bt_music_receiver_simple.ino does not use any ringbuffer! bt_music_receiver_queued.ino however would.

I can't reproduce your issue, so double check the software versions and your cables and/or try with another microcontroller.
And finally make sure that you don't mess up the audio output by logging information.

Also try with a different A2DP source (mobile phone) / application to rule out that the issue is coming from there

**Reply by tnk-rin** [reply]
Oh interesting! I have suspicions that my DACs are the issue as the internal DAC on the ESP worked perfectly.

Thanks for the help!

I did have some success with the queued examples but it still had some audio errors.

**Reply by pschatzmann** [reply]
What DAC are you using ?
To test it just try with a [simple sine output test case](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working).

If the output using AnalogAudioStream is working perfectly, then the issue is really the external DAC...

**Reply by tnk-rin** [reply]
I'm using these PCM5102 modules off of Amazon:  [https://a.co/d/9giCel4](https://a.co/d/9giCel4)

I'll try the sine example you listed and put any updates here.

**Reply by pschatzmann** [reply]
Did you follow the information provided in the [dac documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC) ?

If your solder bridges are not set to the recommended values, you need to make some more connectsions....

**Reply by tnk-rin** [reply]
I did use the External DAC documentation, all of the bridges are correctly wired up. Sine example ended up doing the same thing as the A2DP examples. 

![image](https://github.com/user-attachments/assets/cf1b3ba3-21c8-4068-873b-2754dfcaddf5)

Definitely looks like the DAC is the issue.
